set.mm

```diff
@@ -75869,32 +75869,33 @@ currently used conventions for such cases (see ~ cbvmpt2x , ~ ovmpt2x and
     ${
       $d x A $.
       $( A way of showing an ordinal function is one-to-one.  (Contributed by
-         NM, 9-Feb-1997.) $)
+         NM, 9-Feb-1997.)  Avoid ~ ax-8 .  (Revised by Gino Giotto,
```

Comment:
Can you tell where [bj-ax8](https://us.metamath.org/mpeuni/bj-ax8.html) is used? Metamath.exe doesn't show me:

<pre>Metamath - Version 0.198 7-Aug-2021           Type HELP for help, EXIT to exit.
MM> READ set.mm
Reading source file "set.mm"... 43199123 bytes
43199123 bytes were read into the source buffer.
The source has 205354 statements; 2738 are $a and 40892 are $p.
No errors were found.  However, proofs were not checked.  Type VERIFY PROOF *
if you want to check them.
MM> SHOW TRACE_BACK tz7.48lem /MATCH bj-ax8
The proof of statement "tz7.48lem" uses the following earlier statements:
  (None)
MM></pre>

If true that would be odd since [bj-ax8](https://us.metamath.org/mpeuni/bj-ax8.html) is in a mathbox, and I didn't have `/INCLUDE_MATHBOXES` activated, so it would be a bug of the minimizer.

Also is it even possible to depend from a statement that appears after the main one? I'm new here, so there might be something that I'm missing.

Reply:
I mean that eleq1w is more general than elequ1 but doesn't use ax-8, due to df-clel.
So the way ax-8 is "saved" here is equivalent to the way it is saved in bj-ax8.

Reply:
That's interesting. If [df-clel](https://us.metamath.org/mpeuni/df-clel.html) is too powerful, it seems dangerous to leave it like that, shouldn't it have a `New usage is discouraged` label on it?

However in this case [df-clel](https://us.metamath.org/mpeuni/df-clel.html) was already used in `fvres` and `ordtri3`, so technically it wasn't introduced as a new dependency for `tz7.48lem`. The final outcome is still a clear loss.

Reply:
I originally approved but I did not have @icecream17's sharp eye about axiom dependency!
I think this is pending @icecream17's answer and so I will continue to hold merging for now.

For reference, [there were mailing list exchanges](https://groups.google.com/g/metamath/c/xdHB0oW0aZ4/m/EYUEt3DDBgAJ) between Norm and @benjub about ax-8. I think ultimately, Norm preferred to keep the actual axioms as near to the literature as possible, and so Benoît's suggestions were kept in his mathbox.

@avekens @jkingdon @digama0 don't hesitate to step in if you have an opinion!

Reply:
Hmm, then I'm not sure about what to do with df-clel. I don't think adding `(New usage is discouraged.)` would do much since many theorems using df-clel already avoid ax-8 this way, like eleq1w

Although, here's an idea to automatically verify df-clab, df-cleq, and df-clel:

1. df-clab is the only instance of membership between a setvar and a class, and no other syntax construction is ambiguous to that
2. df-cleq is not the only instance of equality between two classes, specifically there is the equality between two setvars. This can be covered with a hypothesis: `a = b <-> A. x ( x e. a <-> x e. b )`
3. df-clel is not the only instance of membership between two classes, specifically there is the membership between two setvars and df-clab. This can be covered with two hypotheses: `a e. b <-> E. x ( x = a /\ x e. b )` and `a e. { b | ph } <-> E. x ( x = a /\ x e. { b | ph } )`

We also have the setvar distinct variable justifications like with the theoretical cleqjust and (different) cleljust that say `A. x ( x e. A <-> x e. B ) <-> A. y ( y e. A <-> y e. B )` and `E. x ( x = A /\ x e. B ) <-> E. y ( y = A /\ y e. B )`

This supports changing these definitions. (I'm not completely sure on whether to have the universal closure version of these hypotheses like in [bj-df-cleq](https://us.metamath.org/mpeuni/bj-df-cleq.html) and [bj-df-clel](https://us.metamath.org/mpeuni/bj-df-clel.html) though)

Reply:
I am not in favor of using the universial closure here, because this is not the general approach throughout Metamath.

Reply:
> @avekens @jkingdon @digama0 don't hesitate to step in if you have an opinion!

Probably not enough of one to be useful. It would be nice if we could have a bit more settled situation with respect to https://us.metamath.org/mpeuni/df-clel.html and https://us.metamath.org/mpeuni/df-cleq.html but I don't think I know enough about what the choices are and the significance of each to suggest a plan.

Reply:
> It would be nice if we could have a bit more settled situation with respect to https://us.metamath.org/mpeuni/df-clel.html and https://us.metamath.org/mpeuni/df-cleq.html.

I was under the impression that the [mailing group discussion](https://groups.google.com/g/metamath/c/xdHB0oW0aZ4/m/EYUEt3DDBgAJ) settled the point: Norm said he preferred to keep the current axioms, and have the alternative solution in a mathbox, @benjub agreed to do so, and that's the situation we have until now.

Did I understand that correctly? I'm also not diving much into these discussions.

We could reopen that discussion, though.

Reply:
A problem with changing the definition now is that currently cleljust2 would add ax-10 ax-12 ax-13 which obscures their actual usages. These axioms would be removed by the new definition of substitution df-ssb.

So if wanting to introduce ax-8 to df-clel, then the choices are either wait until df-ssb is merged or add ax-8 to df-clel in some arbitrary fashion (like with dfcleq and ax-9)

Reply:
Perhaps one could introduce dfclel which arbitrarily introduces ax-8. This would be a quick solution. I say this now because of #3250 + #3262

Reply:
The ideal of "arbitrarily" introducing ax-8 is, I guess, similar to what we did with ax-9 in https://us.metamath.org/mpeuni/dfcleq.html . But I suppose that sort of thing sidesteps the question of what 
the real status (axiom or definition) of `df-clel` `df-clab` and `df-cleq` is.